enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,972 @@
+   - MicroShift is running for the first time on the device.
+   - There's nothing to backup or migrate, skip to running cluster.
+
+1. If data exists but is missing version metadata:
+   Assume it's 4.13, make a backup (use `4.13` as ID), proceed to [data migration](#data-migration-1).
+
+1. If previous boot was healthy:
+   **backup data for previous boot**, check if next point is applicable, proceed to [data migration](#data-migration-1).
+
+1. (Regardless of previous boot health)<a name="current-commit-different-restore"></a>
+   If currently running commit is different from previous boot's commit
+   (rolled back), and backup exist for current commit:
+   **restore data for current commit**, skip data migration, [start cluster](#starting-the-cluster).
+
+1. If health of previous boot is unknown (e.g. system was unexpectedly rebooted before health check)
+   and commit of current and previous boot is the same:
+   skip to [data migration](#data-migration-1), i.e. retry start up, but check version skews just in case.
+   Otherwise assume it was unhealthy and proceed to the next point.
+
+(Following checks assume that previous boot was unhealthy)
+
+1. If commits of current and previous boots are different, and there's no backup for current commit
+   1. If "history of commits" contains current commit and version metadata matches current boot:
+      **backup data for current commit and [start cluster](#starting-the-cluster)**
+      > It means that current commit was already running on the device.
+      > If the backup is missing, it means that either system was unhealthy or new commit failed to make a backup.
+      > The former is unsupported scenario (system should be healthy or MicroShift data cleaned up),
+      > so only the latter is considered.
+
+   1. If "history of commits" does not contain current commit
+      > Previous boot was unhealthy and it's first boot for current commit
+      > (otherwise it would be in "history of commits")
+      1. Previous boot's commit is not a rollback commit: **delete data and [start cluster](#starting-the-cluster)**.
+         > It means that on actual previous boot didn't run MicroShift.
+         > This fits FIDO scenario.
+
+1. Commits of current and previous boots are the same (current commit booted more than once in a row)
+   1. If backup for current commit exists: **restore**
+      > - If this is greenboot's reboot, then backup was created before this commit was staged and deployed.
+      >   So commit must've been reintroduced and [data would be restored early in the process](#current-commit-different-restore).
+      >   - Being in this place in decision tree means that either restore failed or system was unhealthy.
+      >     Restoring again seems to best bet.
+      >
+      > - If this is manual reboot, then system was healthy,
+      >   then manually rebooted, backed data up and end up unhealthy,
+      >   again rebooted resulting in being in this place in decision tree.
```

Comment:
If the host reboots outside of an ostree/greenboot change, do we take a backup? If not, then we could have a very old backup that is restored when the host reboots on its own (power loss, etc.). That could lead to data loss.

Reply:
If the system was considered healthy by greenboot (close to start of the system), then yes - data will be backed up. greenboot healthchecks run on each boot, but reboots only happens when new deployment is staged (this triggers setting the `boot_counter`).
If the system was healthy, then after backup, it'll consider if prev boot's and current commit are the same (they are in this scenario), and it'll skip all the restores

Reply:
To put it another way: these checks and actions are performed on each boot, regardless if we running commit is different from previous boot (staged / rolled back) or not (reboot)